### PR TITLE
add block split bloom filter

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. use block split bloom filter
2. build bloom filter against data page 
3. add distinct value to bloom filter
4. add ordinal index to bloom filter index

Issue: #1389

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/bloom_filter_index_reader.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+          _bloom_filter_index_meta(bloom_filter_index_meta) {
+        _typeinfo = get_type_info(OLAP_FIELD_TYPE_VARCHAR);
+    }
+
+    Status load();
+
+    // create a new column iterator.
+    Status new_iterator(std::unique_ptr<BloomFilterIndexIterator>* iterator);
+
+    const TypeInfo* type_info() {
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter_index_reader.h]:
```suggestion
    const TypeInfo* type_info() const {
```

File: be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+namespace doris {
+namespace segment_v2 {
+
+const uint32_t BlockSplitBloomFilter::SALT[8] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+        0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+void BlockSplitBloomFilter::add_hash(uint64_t hash) {
+    // most significant 32 bit mod block size as block index(BTW:block size is
+    // power of 2)
+    DCHECK(_size >= BYTES_PER_BLOCK);
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp]: what you want is num_bytes?

File: be/src/olap/rowset/segment_v2/bloom_filter.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    // Factory function for BloomFilter
+    static Status create(BloomFilterAlgorithmPB algorithm, std::unique_ptr<BloomFilter>* bf);
+
+    // Compute the optimal bit number according to the following rule:
+    //     m = -n * ln(fpp) / (ln(2) ^ 2)
+    // n: expected distinct record number
+    // fpp: false positive probablity
+    // the result will be power of 2
+    static uint32_t optimal_bit_num(uint64_t n, double fpp);
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter.h]: why not encapasulate this function in this class. user just pass n to this class.

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -161,6 +161,7 @@ RowsetSharedPtr BetaRowsetWriter::build() {
 OLAPStatus BetaRowsetWriter::_create_segment_writer() {
     auto path = BetaRowset::segment_file_path(_context.rowset_path_prefix, _context.rowset_id, _num_segment);
     segment_v2::SegmentWriterOptions writer_options;
+    writer_options.has_bf_for_unique_agg_column = _context.version.first == 0;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/beta_rowset_writer.cpp]: change it to whether_to_filter_value

File: be/src/olap/rowset/segment_v2/bloom_filter.h [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    // specially add one byte for null flag
+    char* _data;
+    // optimal bloom filter num bytes
+    // it is calculated by optimal_bit_num() / 8
+    uint32_t _num_bytes;
+    // equal to _num_bytes + 1
+    // last byte is for has_null flag
+    uint32_t _size;
+    // last byte's pointer in data for null flag
+    bool* _has_null;
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter.h]: give default value for all these members.

File: be/src/olap/rowset/segment_v2/bloom_filter.h [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    // for write
+    bool init(uint64_t n, double fpp, HashStrategyPB strategy) {
+        if (strategy == HASH_MURMUR3_X64_64) {
+            _hash_func = murmur_hash3_x64_64;
+        } else {
+            return false;
+        }
+        _num_bytes = _optimal_bit_num(n, fpp) / 8;
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter.h]: better to add DCHECK _num_bytes is power of 2

File: be/src/olap/rowset/segment_v2/bloom_filter.h [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    static Status create(BloomFilterAlgorithmPB algorithm, std::unique_ptr<BloomFilter>* bf);
+
+    BloomFilter() = default;
+
+    virtual ~BloomFilter() {
+        delete[] _data;
+    }
+
+    // for write
+    bool init(uint64_t n, double fpp, HashStrategyPB strategy) {
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter.h]: why not return Status?

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 8]
```diff
@@ -224,6 +225,39 @@ Status ColumnReader::_calculate_row_ranges(const std::vector<uint32_t>& page_ind
...
+    std::unique_ptr<BloomFilterIndexIterator> bf_iter;
+    RETURN_IF_ERROR(_bloom_filter_index_reader->new_iterator(&bf_iter));
+    size_t range_size = row_ranges->range_size();
+    // get covered page ids
+    std::set<uint32_t> page_ids;
+    for (int i = 0; i < range_size; ++i) {
+        int64_t from = row_ranges->get_range_from(i);
+        int64_t idx = from;
+        int64_t to = row_ranges->get_range_to(i);
+        auto iter = _ordinal_index->seek_at_or_before(row_ranges->get_range_from(i));
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]:
```suggestion
        auto iter = _ordinal_index->seek_at_or_before(from);
```

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 9]
```diff
@@ -224,6 +225,39 @@ Status ColumnReader::_calculate_row_ranges(const std::vector<uint32_t>& page_ind
...
+        while (idx < to) {
+            page_ids.insert(iter.cur_idx());
+            idx = _ordinal_index->get_last_row_id(iter.cur_idx()) + 1;
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: Why don't get last_row_id from iter?

File: be/src/olap/rowset/segment_v2/bloom_filter.h (hunk 5 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter.h]: added in contructor

File: be/src/olap/rowset/segment_v2/bloom_filter.h (hunk 6 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter.h]: done

File: be/src/olap/rowset/segment_v2/bloom_filter.h (hunk 7 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter.h]: done

File: be/src/olap/rowset/segment_v2/column_reader.cpp (hunk 9 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/column_reader.cpp]: ok, I will add this api in iter.

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**